### PR TITLE
Prevent OoM in a tight retry loop 

### PR DESCRIPTION
**Pull request by basilisk487**
on a machine with very limited resources. 
When the number of push threads is low and there are a lot of items in the retry queue; when a  new batch is also forced to be queued, the retry loop is very likely to immediately acquire back the just released lock, not giving the push thread a chance to write to the queue - and this could last for several minutes, quickly eating up heap memory as points are accumulating in a push thread's buffers.

### CODE REVIEW

**Comment by basilisk487** [comment]
@panghy: Switched from waiting for 50ms to using a less crude approach by using a ReentrantLock in "fair" mode. It also solves the thread starvation issue:
Before:
image

After:
image

File: proxy/src/main/java/com/wavefront/agent/QueuedAgentService.java [hunk 1]
```diff
@@ -451,15 +454,18 @@ public Response postPushData(UUID agentId, UUID workUnitId, Long currentMillis,
   }
 
   private void addTaskToSmallestQueue(ResubmissionTask taskToRetry) {
-    TaskQueue<ResubmissionTask> queue = getSmallestQueue();
+    ResubmissionTaskQueue queue = getSmallestQueue();
     if (queue != null) {
-      synchronized (queue) {
+      try {
```
Comment (panghy) [inline proxy/src/main/java/com/wavefront/agent/QueuedAgentService.java]: these two try blocks can be collapsed

Comment (panghy) [inline proxy/src/main/java/com/wavefront/agent/QueuedAgentService.java]: also, it's normally lock then try ... finally {unlock} (not that it matters i believe in practice but let's say it OOM right after you enter the try block...)

Comment (panghy) [inline proxy/src/main/java/com/wavefront/agent/QueuedAgentService.java]: on second thought, i guess putting it outside the loop can also lock without releasing (depending on how the compiler weaves the code together), anyhow, it's just convention to lock then try :)

File: proxy/src/main/java/com/wavefront/agent/QueuedAgentService.java [hunk 2]
```diff
@@ -169,7 +169,8 @@ public void run() {
               logger.warning("[RETRY THREAD " + threadId + "] TASK STARTING");
             }
             while (taskQueue.size() > 0 && taskQueue.size() > failures) {
-              synchronized (taskQueue) {
+              try {
+                taskQueue.getLockObject().lock();
```
Comment (panghy) [inline proxy/src/main/java/com/wavefront/agent/QueuedAgentService.java]: ditto here for lock then try

**Comment by panghy** [comment]
Thanks for looking into this

**Comment by basilisk487** [comment]
@panghy Done!

**Pull request merged**